### PR TITLE
Register readable prefix for grouped keymap settings

### PR DESCRIPTION
**Pull request by Cyberczy**

### CODE REVIEW

**Comment by ayamir** [comment]
IMO it's better to register prefix for `which-key` when the keymappings are binded, i.e., invoking the `bind.nvim_load_mapping`.

**Comment by Cyberczy** [comment]
It's up to you. I'm just making a suggestion.

**Comment by Cheny-chui** [comment]
I have made some improvements to this pr based on ayamir's suggestion, can you please give me modification access to this pr? @Cyberczy (I'm not sure if I need your action, but I really can't push to your repository). Or maybe need the maintainer's action @ayamir .

**Comment by Cyberczy** [comment]
> 我根据ayamir的建议对这个pr做了一些改进，你能给我这个pr的修改权限吗？@Cyberczy（我不确定是否需要您的操作，但我确实无法自动到达您的存储库）。或者可能维护者的操作@ayamir。

I don't have the authority, lol

**Comment by ayamir** [comment]
It seems better that @Cyberczy give the authority of your fork to @Cheny-chui as the GitHub Docs says: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/committing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by Cyberczy** [comment]
Ok, I've just given you modification access. @Cheny-chui

**Comment by ayamir** [comment]
@CharlesChiuGit Could you please review this PR?

**Comment by CharlesChiuGit** [comment]
ok, let me take a look.

File: lazy-lock.json
Comment (CharlesChiuGit) [inline lazy-lock.json:1]: please remove lazy-lock

**Review by CharlesChiuGit — CHANGES_REQUESTED** [review]
plz remove lazy-lock

File: lazy-lock.json
Comment (Cyberczy) [reply lazy-lock.json:1]: ok

**Comment by ayamir** [comment]
Current prefix settings is not enough and need to be added manually. I'm thinking automatically setting whether is a better idea. Though this requires keymap setup to keep align with it.

Additionally, maybe we can refactor current keymap settings through this PR, such as:
`<leader>D` -> `<leader>hd`
`<leader><leader>D` -> `<leader>hD`
What do you think? @CharlesChiuGit @Jint-lzxy

**Comment by vabatta** [comment]
What is the general rationale behind the current key mapping?
Is it quick commands, frequent commands or "grouped" commands (like [LazyVim](https://www.lazyvim.org/keymaps) which maps into groups like `<leader>w` for windows, `<leader>b` for buffers, ...)?

**Review by CharlesChiuGit — DISMISSED** [review]

**Comment by ayamir** [comment]
> What is the general rationale behind the current key mapping? Is it quick commands, frequent commands or "grouped" commands (like [LazyVim](https://www.lazyvim.org/keymaps) which maps into groups like `<leader>w` for windows, `<leader>b` for buffers, ...)?

Current keymap settings are mixed but a little confusing due to history reasons.

**Comment by CharlesChiuGit** [comment]
```lua
--- Need to expand by hand
local prefix_desc = {
	["<leader>b"] = icons.ui_sep.Buffer .. "Buffer",
	["<leader>c"] = icons.ui_sep.Character .. "Character",
	["<leader>d"] = icons.ui_sep.Bug .. "Debug",
	["<leader>f"] = icons.ui_sep.Telescope .. "Fuzzy Find",
	["<leader>h"] = icons.ui_sep.Git .. "Git Hunk",
	["<leader>l"] = icons.ui_sep.Lsp .. "Lsp",
	["<leader>p"] = icons.ui_sep.Package .. "Package",
	["<leader>n"] = icons.ui_sep.FolderOpen .. "Nvim Tree",
	["<leader>s"] = icons.ui_sep.Tmux .. "Session",
	["<leader>t"] = icons.ui_sep.Lsp .. "Lsp",
	["<leader><space>"] = icons.ui_sep.Git .. "Git: Close diff",
}
```
After pressing leader key one time, some prefix still can be shown, hmmm.
image

**Comment by Cyberczy** [comment]
> `lua\keymap\prefix.lua`
> 
> ```lua
> --- Need to expand by hand
> local prefix_desc = {
> 	["<leader>b"] = icons.ui_sep.Buffer .. "Buffer",
> 	["<leader>c"] = icons.ui_sep.Character .. "Character",
> 	["<leader>d"] = icons.ui_sep.Bug .. "Debug",
> 	["<leader>f"] = icons.ui_sep.Telescope .. "Fuzzy Find",
> 	["<leader>h"] = icons.ui_sep.Git .. "Git Hunk",
> 	["<leader>l"] = icons.ui_sep.Lsp .. "Lsp",
> 	["<leader>p"] = icons.ui_sep.Package .. "Package",
> 	["<leader>n"] = icons.ui_sep.FolderOpen .. "Nvim Tree",
> 	["<leader>s"] = icons.ui_sep.Tmux .. "Session",
> 	["<leader>t"] = icons.ui_sep.Lsp .. "Lsp",
> 	["<leader><space>"] = icons.ui_sep.Git .. "Git: Close diff",
> }
> ```
> 
> 按一次leader键后，仍然可以显示一些前缀，嗯。 图像


`<leader><space>` seems to have no effect.

**Comment by CharlesChiuGit** [comment]
> `<leader><space>` seems to have no effect.

yeah, im not sure how to set it. i have tried `<leader><space>` or `<leader><leader>` both r not working.
also, `<leader>h` and `<leader>l` r not valid too.

**Comment by CharlesChiuGit** [comment]
@ayamir @Jint-lzxy 
since our keymap group is a bit confusing right now, our which-key prefix nerdfont hit might be a bit misleading too.

**Comment by ayamir** [comment]
IMO it's better to refactor it.

**Comment by ayamir** [comment]
It looks tidy now.
image

**Comment by Cyberczy** [comment]
LGTM!!!

**Comment by Cyberczy** [comment]
The right half of which-key seems a bit empty. Perhaps we can add a bit of padding.

**Comment by ayamir** [comment]
> The right half of which-key seems a bit empty. Perhaps we can add a bit of padding.

IMO it is unnecessary.

**Comment by Cyberczy** [comment]
> > The right half of which-key seems a bit empty. Perhaps we can add a bit of padding.
> 
> IMO it is unnecessary.

Alright, it's just a suggestion

**Comment by Jint-lzxy** [comment]
Sorry for the delay! My sched's been a bit tight recently. But I'll probably review this in the next day or two lol

**Review by CharlesChiuGit — APPROVED** [review]
LGTM

**Review by Jint-lzxy — APPROVED** [review]
LGTM! This PR is awesome lmao
***
(idk if this is related) But I'm wondering if the real issue might lie in the implementation of [which-key.nvim](https://github.com/folke/which-key.nvim), as I used a [different approach](https://github.com/Jint-lzxy/nvimconfig/blob/f5e4e72cb7ab884c7afef4274c64a991ad850713/lua/modules/configs/tool/which-key.lua#L33) to handle keymap registration, but it seems that (only recently) which-key sometimes displays `+prefix` even for named groups (like `<leader>h`, see the screenshot below):
S

Then I did a git bisect, and the issue seems to originate from [this commit](https://github.com/folke/which-key.nvim/commit/8503c0d725420b37ac31e44753657cde91435597) (strangely tho, this means that which-key thinks some of my keymap groups don't have a valid name, but obviously I _did_ give them names). For some reason, in [this call](https://github.com/folke/which-key.nvim/blob/4433e5ec9a507e5097571ed55c02ea9658fb268a/lua/which-key/keys.lua#L116), the (to be processed) group still has a name, but immediately [in the next line](https://github.com/folke/which-key.nvim/blob/4433e5ec9a507e5097571ed55c02ea9658fb268a/lua/which-key/keys.lua#L117), that name disappears. Based on this, IMO a simpler solution 'might' be to revert [`8503c0`](https://github.com/folke/which-key.nvim/commit/8503c0d725420b37ac31e44753657cde91435597) (I can create a fork just like we made a temporary fix for `lsp-signature`), and then wait for folke's response (I'll open an issue for this); or just merge this PR. What do you guys think @ayamir @CharlesChiuGit?

**Comment by ayamir** [comment]
Made improvement for startup time.
Before: 
image
Now:
image

**Comment by ayamir** [comment]
> (idk if this is related) But I'm wondering if the real issue might lie in the implementation of [which-key.nvim](https://github.com/folke/which-key.nvim), as I used a [different approach](https://github.com/Jint-lzxy/nvimconfig/blob/f5e4e72cb7ab884c7afef4274c64a991ad850713/lua/modules/configs/tool/which-key.lua#L33) to handle keymap registration, but it seems that (only recently) which-key sometimes displays `+prefix` even for named groups (like `<leader>h`, see the screenshot below): S

I didn't view settings about `<leader>h` for group name in https://github.com/Jint-lzxy/nvimconfig/blob/f5e4e72cb7ab884c7afef4274c64a991ad850713/lua/modules/configs/tool/which-key.lua#L34-L145

**Comment by Jint-lzxy** [comment]
> I didn't view settings about `<leader>h` for group name ...

Ah, I think this probably isn't related. Sorry for the noise here! (I didn't notice that our impl is kinda like an event loop 🙈) Here's a similar issue with more details: https://github.com/folke/which-key.nvim/issues/482.

@ayamir With that being said, I'm still a bit puzzled as to why we need to register these mappings dynamically. I feel like just statically registering them using `require("which-key").register({})` would be enough, since `which-key` should do the grouping work (correct me if I'm misunderstanding something lmao). Registering a callback that gets invoked every 300 milliseconds seems a bit inefficient to me.

**Comment by ayamir** [comment]
> With that being said, I'm still a bit puzzled as to why we need to register these mappings dynamically. I feel like just statically registering them using require("which-key").register({}) would be enough, since which-key should do the grouping work (correct me if I'm misunderstanding something lmao).

It's b/c the statical way only register the group name once when init and doesn't work for our lsp-related keymaps are binded with buffer after `LspAttach` event. But after the keymap refactored, this problem is addressed now (b/c `trouble.nvim`-related keymaps were merged into the `<leader>l` group). Indeed, now we can use the statical way to avoid unnecessary complexity lol.
image

**Comment by Cyberczy** [comment]
So, should we stick with statically registering them, or should we maintain the current state by dynamically registering them?

**Review by ayamir — APPROVED** [review]

**Pull request merged**